internal/graph/dispatch_throttling_check_resolver_test.go

```diff
@@ -0,0 +1,105 @@
+package graph
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/goleak"
+)
+
+type countResolveCheck struct {
+	count *atomic.Uint32
+}
+
+func newCountResolveCheck() *countResolveCheck {
+	return &countResolveCheck{
+		count: &atomic.Uint32{},
+	}
+}
+
+func (c *countResolveCheck) ResolveCheck(ctx context.Context, req *ResolveCheckRequest) (*ResolveCheckResponse, error) {
+	c.count.Add(1)
+	return &ResolveCheckResponse{Allowed: true}, nil
+}
+
+func (c *countResolveCheck) Close() {}
+
+func (c *countResolveCheck) NumResolveCheck() uint32 {
+	return c.count.Load()
+}
+
+func TestDispatchThrottlingCheckResolver(t *testing.T) {
+	t.Cleanup(func() {
+		goleak.VerifyNone(t)
+	})
+	t.Run("below_shaping_level_should_not_be_shaped", func(t *testing.T) {
+		dispatchThrottlingCheckResolverConfig := DispatchThrottlingCheckResolverConfig{
+			TimerTickerFrequency: 1 * time.Hour,
+			Level:                200,
+			Rate:                 8,
+		}
+		dut := NewDispatchThrottlingCheckResolver(dispatchThrottlingCheckResolverConfig)
+		defer dut.Close()
+
+		c := newCountResolveCheck()
+		dut.SetDelegate(c)
+
+		dispatchCounter := &atomic.Uint32{}
+		_, err := dut.ResolveCheck(context.Background(), &ResolveCheckRequest{DispatchCounter: dispatchCounter})
+		require.NoError(t, err)
+		require.Equal(t, uint32(1), c.NumResolveCheck())
+		require.Equal(t, uint32(1), dispatchCounter.Load())
+	})
+
+	t.Run("above low shaping level should be shaped very aggressively", func(t *testing.T) {
+		dispatchThrottlingCheckResolverConfig := DispatchThrottlingCheckResolverConfig{
+			TimerTickerFrequency: 1 * time.Hour,
+			Level:                200,
+			Rate:                 2,
+		}
+		dut := NewDispatchThrottlingCheckResolver(dispatchThrottlingCheckResolverConfig)
+		defer dut.Close()
+
+		c := newCountResolveCheck()
+		dut.SetDelegate(c)
+
+		dispatchCounter := &atomic.Uint32{}
+		dispatchCounter.Store(201)
+
+		var goFuncDone sync.WaitGroup
+		goFuncDone.Add(1)
+
+		var goFuncInitiated sync.WaitGroup
+		goFuncInitiated.Add(1)
+
+		go func() {
+			defer goFuncDone.Done()
+			goFuncInitiated.Done()
+			_, err := dut.ResolveCheck(context.Background(), &ResolveCheckRequest{DispatchCounter: dispatchCounter})
+			//nolint:testifylint
+			require.NoError(t, err)
+			//nolint:testifylint
+			require.Equal(t, uint32(202), dispatchCounter.Load())
+		}()
+
+		goFuncInitiated.Wait()
+		time.Sleep(1 * time.Millisecond)
+		require.Equal(t, uint32(0), c.NumResolveCheck())
+
+		// the first time tick will not release the job
+		newCount := dut.handleTimeTick(0)
+		time.Sleep(1 * time.Millisecond)
+		require.Equal(t, uint32(0), c.NumResolveCheck())
+		require.Equal(t, uint32(1), newCount)
+
+		// the second time tick will release the job
+		newCount = dut.handleTimeTick(newCount)
+		goFuncDone.Wait()
+		require.Equal(t, uint32(1), c.NumResolveCheck())
```

Comment:
This test is confusing at first glance, because if the tick frequency is 1hr and we require 2 ticks, then I wouldn't expect ResolveCheck to increment the dispatch count until 2h later.

Reply:
added comment to clarify